Unable to use matrix view with Eigen::Ref and Eigen::Map

If I try to compile the these lines 
```cpp
     iDynTree::MatrixDynSize a;
     m_kinDyn->getFrameFreeFloatingJacobian(m_frameIndex,
                                            iDynTree::toEigen(a));
```
the following error is thrown by the compiler
```
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/KinDynComputations.h:697:10: note:   no known conversion for argument 2 from ‘Eigen::Map<Eigen::Matrix<double, -1, -1, 1> >’ to ‘iDynTree::MatrixView<double>’
```


Something similar happens if I do

```cpp
     m_kinDyn->getFrameFreeFloatingJacobian(m_frameIndex,
                                            this->subA(m_robotAccelerationVariable));

```

where the function `subA` is
```cpp
Eigen::Ref<Eigen::MatrixXd> subA(const VariablesHandler::VariableDescription& description)
{
    return m_A.block(0, description.offset, 6, description.size);
}
```

This is the error
```
bool iDynTree::KinDynComputations::getFrameFreeFloatingJacobian(iDynTree::FrameIndex, iDynTree::MatrixView<double>)
     bool getFrameFreeFloatingJacobian(const FrameIndex frameIndex,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/KinDynComputations.h:697:10: note:   no known conversion for argument 2 from ‘Eigen::Ref<Eigen::Matrix<double, -1, -1> >’ to ‘iDynTree::MatrixView<double>’
```

Actually supporting the eigen::map is not really important since matrixview can be seen as a map, so we can wrap directly a non-idyntree object using the matrixview. On the other hand, adding the support to `eigen::ref` it would be really nice since it may happen that the input of a function is an `Eigen::Ref` object. 
I think that the problem can be split into these subtasks:
1. understand why the eigen::ref doesn't work out of the box.
2. implement a `SubMatrixView`. This can be used to extract a block from a `MatrixView` (this is actually what I need) 

cc @traversaro and @S-Dafarra

In order to replicate the error, try to instantiate a `MatrixVew` from an `eigenRef`, i.e. 

```cpp
     iDynTree::MatrixView a(eigenRef);
```

```
/home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:88:36: error: class template argument deduction failed:
     iDynTree::MatrixView a(eigenRef);
                                    ^
/home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:88:36: error: no matching function for call to ‘MatrixView(Eigen::Ref<Eigen::Matrix<double, -1, -1> >&)’
In file included from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixDynSize.h:17:0,
                 from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/KinDynComputations.h:18,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/OptimalControlElement.h:15,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/SE3Task.h:11,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:8:
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:169:9: note: candidate: template<class ElementType> MatrixView(iDynTree::MatrixView<ElementType>::pointer, iDynTree::MatrixView<ElementType>::index_type, iDynTree::MatrixView<ElementType>::index_type, const iDynTree::MatrixStorageOrdering&)-> iDynTree::MatrixView<ElementType>
         MatrixView(pointer in_data,
         ^~~~~~~~~~
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:169:9: note:   template argument deduction/substitution failed:
/home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:88:36: note:   mismatched types ‘iDynTree::MatrixView<ElementType>::element_type*’ and ‘Eigen::Ref<Eigen::Matrix<double, -1, -1> >’
     iDynTree::MatrixView a(eigenRef);
                                    ^
In file included from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixDynSize.h:17:0,
                 from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/KinDynComputations.h:18,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/OptimalControlElement.h:15,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/SE3Task.h:11,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:8:
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:162:9: note: candidate: template<class ElementType, class Container, typename std::enable_if<((std::is_convertible<decltype (declval<Container>().data()), ElementType*>::value && (! iDynTree::MatrixViewInternal::has_IsRowMajor<Container, void>::value)) && (! std::is_same<Container, iDynTree::MatrixView<ElementType> >::value)), int>::type <anonymous> > MatrixView(Container&, const iDynTree::MatrixStorageOrdering&)-> iDynTree::MatrixView<ElementType>
         MatrixView(Container& matrix, const MatrixStorageOrdering& order = MatrixStorageOrdering::RowMajor)
         ^~~~~~~~~~
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:162:9: note:   template argument deduction/substitution failed:
/home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:88:36: note:   couldn't deduce template parameter ‘ElementType’
     iDynTree::MatrixView a(eigenRef);
                                    ^
In file included from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixDynSize.h:17:0,
                 from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/KinDynComputations.h:18,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/OptimalControlElement.h:15,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/SE3Task.h:11,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:8:
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:147:9: note: candidate: template<class ElementType, class Container, typename std::enable_if<((std::is_convertible<decltype (declval<Container>().data()), ElementType*>::value && iDynTree::MatrixViewInternal::has_IsRowMajor<Container, void>::value) && (! std::is_same<Container, iDynTree::MatrixView<ElementType> >::value)), int>::type <anonymous> > MatrixView(Container&)-> iDynTree::MatrixView<ElementType>
         MatrixView(Container& matrix)
         ^~~~~~~~~~
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:147:9: note:   template argument deduction/substitution failed:
/home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:88:36: note:   couldn't deduce template parameter ‘ElementType’
     iDynTree::MatrixView a(eigenRef);
                                    ^
In file included from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixDynSize.h:17:0,
                 from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/KinDynComputations.h:18,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/OptimalControlElement.h:15,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/SE3Task.h:11,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:8:
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:136:9: note: candidate: template<class ElementType, class Container, typename std::enable_if<(((std::is_const<_Tp>::value && std::is_convertible<decltype (declval<Container>().data()), ElementType*>::value) && (! iDynTree::MatrixViewInternal::has_IsRowMajor<Container, void>::value)) && (! std::is_same<Container, iDynTree::MatrixView<ElementType> >::value)), int>::type <anonymous> > MatrixView(const Container&, const iDynTree::MatrixStorageOrdering&)-> iDynTree::MatrixView<ElementType>
         MatrixView(const Container& matrix, const MatrixStorageOrdering& order = MatrixStorageOrdering::RowMajor)
         ^~~~~~~~~~
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:136:9: note:   template argument deduction/substitution failed:
/home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:88:36: note:   couldn't deduce template parameter ‘ElementType’
     iDynTree::MatrixView a(eigenRef);
                                    ^
In file included from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixDynSize.h:17:0,
                 from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/KinDynComputations.h:18,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/OptimalControlElement.h:15,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/SE3Task.h:11,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:8:
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:119:9: note: candidate: template<class ElementType, class Container, typename std::enable_if<(((std::is_const<_Tp>::value && std::is_convertible<decltype (declval<Container>().data()), ElementType*>::value) && iDynTree::MatrixViewInternal::has_IsRowMajor<Container, void>::value) && (! std::is_same<Container, iDynTree::MatrixView<ElementType> >::value)), int>::type <anonymous> > MatrixView(const Container&)-> iDynTree::MatrixView<ElementType>
         MatrixView(const Container& matrix)
         ^~~~~~~~~~
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:119:9: note:   template argument deduction/substitution failed:
/home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:88:36: note:   couldn't deduce template parameter ‘ElementType’
     iDynTree::MatrixView a(eigenRef);
                                    ^
In file included from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixDynSize.h:17:0,
                 from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/KinDynComputations.h:18,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/OptimalControlElement.h:15,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/SE3Task.h:11,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:8:
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:106:28: note: candidate: template<class ElementType, class OtherElementType, class> MatrixView(const iDynTree::MatrixView<OtherElementType>&)-> iDynTree::MatrixView<ElementType>
         IDYNTREE_CONSTEXPR MatrixView(const MatrixView<OtherElementType>& other)
                            ^~~~~~~~~~
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:106:28: note:   template argument deduction/substitution failed:
/home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:88:36: note:   ‘Eigen::Ref<Eigen::Matrix<double, -1, -1> >’ is not derived from ‘const iDynTree::MatrixView<OtherElementType>’
     iDynTree::MatrixView a(eigenRef);
                                    ^
In file included from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixDynSize.h:17:0,
                 from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/KinDynComputations.h:18,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/OptimalControlElement.h:15,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/SE3Task.h:11,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:8:
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:96:9: note: candidate: template<class ElementType> MatrixView(const iDynTree::MatrixView<ElementType>&)-> iDynTree::MatrixView<ElementType>
         MatrixView(const MatrixView& other)
         ^~~~~~~~~~
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:96:9: note:   template argument deduction/substitution failed:
/home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:88:36: note:   ‘Eigen::Ref<Eigen::Matrix<double, -1, -1> >’ is not derived from ‘const iDynTree::MatrixView<ElementType>’
     iDynTree::MatrixView a(eigenRef);
                                    ^
In file included from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixDynSize.h:17:0,
                 from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/KinDynComputations.h:18,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/OptimalControlElement.h:15,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/SE3Task.h:11,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:8:
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:93:9: note: candidate: template<class ElementType> MatrixView()-> iDynTree::MatrixView<ElementType>
         MatrixView()
         ^~~~~~~~~~
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:93:9: note:   template argument deduction/substitution failed:
/home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:88:36: note:   candidate expects 0 arguments, 1 provided
     iDynTree::MatrixView a(eigenRef);
                                    ^
In file included from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixDynSize.h:17:0,
                 from /home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/KinDynComputations.h:18,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/OptimalControlElement.h:15,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/include/BipedalLocomotion/TSID/SE3Task.h:11,
                 from /home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:8:
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:64:40: note: candidate: template<class ElementType> MatrixView(iDynTree::MatrixView<ElementType>)-> iDynTree::MatrixView<ElementType>
     template <class ElementType> class MatrixView
                                        ^~~~~~~~~~
/home/gromualdi/robot-code/robotology-superbuild/build/install/include/iDynTree/Core/MatrixView.h:64:40: note:   template argument deduction/substitution failed:
/home/gromualdi/robot-code/bipedal-locomotion-framework/src/TSID/src/SE3Task.cpp:88:36: note:   ‘Eigen::Ref<Eigen::Matrix<double, -1, -1> >’ is not derived from ‘iDynTree::MatrixView<ElementType>’
     iDynTree::MatrixView a(eigenRef);
                                    ^
src/TSID/CMakeFiles/TSID.dir/build.make:94: recipe for target 'src/TSID/CMakeFiles/TSID.dir/src/SE3Task.cpp.o' failed
make[2]: *** [src/TSID/CMakeFiles/TSID.dir/src/SE3Task.cpp.o] Error 1
CMakeFiles/Makefile2:2706: recipe for target 'src/TSID/CMakeFiles/TSID.dir/all' failed
make[1]: *** [src/TSID/CMakeFiles/TSID.dir/all] Error 2
Makefile:159: recipe for target 'all' failed
make: *** [all] Error 2
```

Today I had the opportunity to understand deeper the problem.

First of all the problem that I underlined in https://github.com/robotology/idyntree/issues/797#issuecomment-750630820
```cpp
iDynTree::MatrixView a(eigenRef);
```
it is simply a template deduction problem (as @S-Dafarra suggested in https://github.com/robotology/idyntree/issues/797#issuecomment-754675295). indeed `iDynTree::MatrixView<double> a(eigenRef)` compiles without any problem.

For instance this snippet
```cpp
#include <iostream>
#include <Eigen/Dense>
#include <iDynTree/Core/MatrixView.h>

void foo(iDynTree::MatrixView<double> view) {
  std::cout << "Size: " << view.rows() << " x " << view.cols() << std::endl;
  std::cout << "view(0,0): " << view(0, 0) << std::endl;
  view(0, 0) = 2;
  std::cout << "view(0,0): " << view(0, 0) << std::endl;
}

int main() {
  Eigen::MatrixXd m(10, 10);
  Eigen::MatrixXd n(10, 10);
  m.setZero();
  n.setIdentity();
  Eigen::Ref<Eigen::MatrixXd> ref(m);

  std::cout << "Eigen::Ref" << std::endl;
  foo(ref);

  std::cout << std::endl;
  std::cout << "Eigen::MatrixXd" << std::endl;
  foo(n);

  return 0;
}
```
gives 
```
Eigen::Ref
Size: 10 x 10
view(0,0): 0
view(0,0): 2

Eigen::MatrixXd
Size: 10 x 10
view(0,0): 1
view(0,0): 2
```

On the other hand as @traversaro and @S-Dafarra explained the `Eigen::Ref` is able to handle also a non-contiguous area of memory. (A block of a dense matrix is represented by a non-contiguous memory). Fortunately, if you try to pass  matrix sublock to `foo` the following error is thrown by the compiler (gcc 10)
```
/home/gromualdi/robot-code/test_eigen/span_eigen.cpp: In function ‘int main()’:
/home/gromualdi/robot-code/test_eigen/span_eigen.cpp:32:14: error: could not convert ‘((Eigen::DenseBase<Eigen::Matrix<double, -1, -1> >*)(& n))->Eigen::DenseBase<Eigen::Matrix<double, -1, -1> >::block(1, 1, 1, 1)’ from ‘Eigen::DenseBase<Eigen::Matrix<double, -1, -1> >::BlockXpr’ {aka ‘Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, -1, false>’} to ‘iDynTree::MatrixView<double>’
   32 |   foo(n.block(1,1,1,1));
      |       ~~~~~~~^~~~~~~~~
      |              |
      |              Eigen::DenseBase<Eigen::Matrix<double, -1, -1> >::BlockXpr {aka Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, -1, false>}
``` 

This problem does not subsist if `foo` takes as input an `iDynTree::MatrixView<const double>`. Indeed in this case `foo(n.block(1,1,1,1));` compiles without any issue. This happens because a temporary object is created at runtime.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,6 +9,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 - Add the possibility to plot and update frames in the Matlab visualizer.
+- Add the possibility to extract submatrix with MatrixView (https://github.com/robotology/idyntree/pull/800)
 
 ## [2.0.2] - 2020-12-04
 

--- a/src/core/include/iDynTree/Core/MatrixView.h
+++ b/src/core/include/iDynTree/Core/MatrixView.h
@@ -77,14 +77,17 @@ namespace iDynTree
 
         MatrixStorageOrdering m_storageOrder;
 
+        index_type m_innerStride;
+        index_type m_outerStride;
+
         index_type rawIndex(index_type row, index_type col) const
         {
             if (m_storageOrder == MatrixStorageOrdering::RowMajor)
             {
-                return (col + this->m_cols * row);
+                return (col + this->m_outerStride * row);
             } else
             {
-                return (this->m_rows * col + row);
+                return (this->m_outerStride * col + row);
             }
         }
 
@@ -174,6 +177,8 @@ namespace iDynTree
             , m_rows(in_rows)
             , m_cols(in_cols)
             , m_storageOrder(order)
+            , m_innerStride(1)
+            , m_outerStride(order == MatrixStorageOrdering::RowMajor ? in_cols : in_rows)
         {
         }
 
@@ -210,6 +215,25 @@ namespace iDynTree
             return this->m_cols;
         }
         ///@}
+
+        MatrixView block(index_type startingRow, index_type startingColumn,
+                         index_type rows, index_type cols) const
+        {
+            assert(rows <= this->rows());
+            assert(cols <= this->cols());
+
+            const index_type offset = rawIndex(startingRow, startingColumn);
+            assert(offset < this->rows() * this->cols());
+
+            MatrixView block;
+            block = *this;
+            block.m_rows = rows;
+            block.m_cols = cols;
+            block.m_storage = this->m_storage + offset;
+
+            return block;
+        }
+
     };
 
     template <class ElementType>

--- a/src/core/tests/MatrixViewUnitTest.cpp
+++ b/src/core/tests/MatrixViewUnitTest.cpp
@@ -29,27 +29,41 @@ int main()
     // test with iDynTree matrix
     MatrixDynSize mat1(27, 41);
     getRandomMatrix(mat1);
-    MatrixView<double> view1(mat1);
+    MatrixDynSize subMat1(10, 23);
+    toEigen(subMat1) = toEigen(mat1).block(3, 5, 10, 23);
 
+    MatrixView<double> view1(mat1);
     areMatricesEqual(view1, mat1);
+
+    auto subView1 = view1.block(3, 5, 10, 23);
+    areMatricesEqual(subView1, subMat1);
 
     // Test with eigen column major matrix
     Eigen::MatrixXd mat2(27, 41);
     mat2.setRandom();
-    MatrixView<double> view2(mat2);
+    Eigen::MatrixXd subMat2;
+    subMat2 = mat2.block(3, 5, 10, 23);
 
+    MatrixView<double> view2(mat2);
     areMatricesEqual(view2, mat2);
+
+    auto subView2 = view2.block(3, 5, 10, 23);
+    areMatricesEqual(subView2, subMat2);
 
     // Test with eigen row major matrix
     Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> mat3(27, 41);
     mat3.setRandom();
-    MatrixView<double> view3(mat3);
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> subMat3;
+    subMat3 = mat3.block(3, 5, 10, 23);
 
+    MatrixView<double> view3(mat3);
     areMatricesEqual(view3, mat3);
 
     auto view4 = make_matrix_view(mat3);
     areMatricesEqual(view4, mat3);
 
+    auto subView3 = view3.block(3, 5, 10, 23);
+    areMatricesEqual(subView3, subMat3);
 
     return 0;
 }

